src/Hamiltonians/geometry.jl

```diff
@@ -1,256 +1,227 @@
-PeriodicBoundaries(5, 4)
+Return the number of dimensions of the lattice in this geometry.
+"""
+num_dimensions(::Geometry{D}) where {D} = D
 
-julia> num_neighbours(lattice)
-4
+"""
+    fold_vec(g::Geometry{D}, vec::SVector{D,Int}) -> SVector{D,Int}
 
-julia> neighbour_site(lattice, 1, 1)
-2
+Use the Geometry to fold the `vec` in each dimension. If folding is disabled in a
+dimension, and the vector is allowed to go out of bounds.
 
-julia> neighbour_site(lattice, 1, 2)
-5
+```julia
+julia> geo = Geometry((2,3), (true,false))
+Geometry{2}((2, 3), (true, false))
 
-julia> neighbour_site(lattice, 1, 3)
-6
+julia> fold_vec(geo, (3,1))
+(1, 1)
 
-julia> neighbour_site(lattice, 1, 4)
-16
+julia> fold_vec(geo, (3,4))
+(1, 4)
 ```
-
-## See also
-
-* [`LatticeGeometry`](@ref)
-* [`HardwallBoundaries`](@ref)
-* [`LadderBoundaries`](@ref)
-* [`num_neighbours`](@ref)
-* [`neighbour_site`](@ref)
 """
-struct PeriodicBoundaries{D} <: LatticeGeometry{D}
-    size::NTuple{D,Int}
+function fold_vec(g::Geometry{D}, vec::SVector{D,Int}) where {D}
+    (_fold_vec(Tuple(vec), fold(g), size(g)))
+end
+@inline _fold_vec(::Tuple{}, ::Tuple{}, ::Tuple{}) = ()
+@inline function _fold_vec((x, xs...), (f, fs...), (d, ds...))
+    x = f ? mod1(x, d) : x
+    return (x, _fold_vec(xs, fs, ds)...)
 end
 
-PeriodicBoundaries(args::Vararg{Int}) = PeriodicBoundaries(args)
-
-Base.size(geom::PeriodicBoundaries) = geom.size
-
-num_neighbours(::PeriodicBoundaries{D}) where {D} = 2D
-
-function neighbour_site(geom::PeriodicBoundaries{D}, site, i) where {D}
-    # Neighbour indexing
-    # i |  x  y  z …
-    # 0 | +1  0  0
-    # 1 | -1  0  0
-    # 2 |  0 +1  0
-    # 3 |  0 -1  0
-    # 4 |  0  0 +1
-    #         ⋮    ⋱
-    i -= 1
-    cart_indices = CartesianIndices(size(geom))
-    cart_index = Tuple(cart_indices[site])
-    offset = ntuple(Val(D)) do k
-        ifelse(2(k - 1) ≤ i < 2k, ifelse(iseven(i), 1, -1), 0)
-    end
-    new_index = CartesianIndex(mod1.(cart_index .+ offset, size(geom) .% UInt))
-    return LinearIndices(cart_indices)[new_index]
+function Base.getindex(g::Geometry{D}, vec::Union{NTuple{D,Int},SVector{D,Int}}) where {D}
+    return get(LinearIndices(size(g)), fold_vec(g, SVector(vec)), 0)
 end
+Base.getindex(g::Geometry, i::Int) = SVector(Tuple(CartesianIndices(size(g))[i]))
 
 """
-    HardwallBoundaries
-
-Rectangular lattice with hard wall boundary conditions of size `size`.
-[`neighbour_site()`](@ref) will return 0 for some neighbours of boundary sites.
+    UnitVectors(D) <: AbstractVector{SVector{D,Int}}
+    UnitVectors(geometry::Geometry) <: AbstractVector{SVector{D,Int}}
 
-The dimension of the lattice is controlled by the number of arguments given to its
-constructor.
+Iterate over unit vectors in `D` dimensions.
 
-## Example
+```jldoctest; setup=:(using Rimu.Hamiltonians: UnitVectors)
+julia> UnitVectors(3)
+6-element UnitVectors{3}:
+ [1, 0, 0]
+ [0, 1, 0]
+ [0, 0, 1]
+ [-1, 0, 0]
+ [0, -1, 0]
+ [0, 0, -1]
```

Comment:
Hmm, shouldn't there be only three unit vectors for a three-dimensional space? If this is useful as it maybe it needs a different name and more verbose explanation.

Please add a hyperlink back to `Geometry`.